[libraries/Utility] Add startsWith, endsWith string functions to Utility (with example)

johnsaigle opened the pull request:
- [x] Create syntax sugar for `substr` to prevent developer errors with string comparison

- Add startsWith string comparison function to Utility

- Add endsWith string comparison function to Utility

- see #2966 for an example of this error

driusan commented on php/libraries/Utility.class.inc:
```diff
@@ -1277,5 +1277,32 @@ class Utility
+    static function endsWith($haystack, $needle)
+    {
+        return substr($haystack, -strlen($needle)) === $needle;
+    }
```
I think these would make more sense in a new Strings class, rather than adding more stuff to the already overloaded Utility class.

ridz1208 commented:
@johnsaigle @driusan I think the current way these are implemented is problematic for a couple reasons:

 - with the current hectic state of the utility class, these functions will get lost in the stack and noone will use someone goes and replaces all substr calls to use these.. I'm scared they will find their way yo a PR like https://github.com/aces/Loris/pull/2913  very very quickly

 - I think its fair to expect a developer to know how to find/read documentation on php function usage... personally I feel like this will make developers more lazy !!


I wouldn't be opposed to create a new String.class.inc and have those there (they will be easier to find) the same PR should have functions such as contains(needle,haystack) and postion(needle,haystack) But i would insist on these being used multiple places in the code before the PR is merged

johnsaigle commented:
@ridz1208 I am in favour of a Strings class to make things simpler. I added these not so much because of new developers but because of PHP being pretty inconsistent. 

In fact I added them because of #2953. Misuse of `strpos` caused some critical security issues and (not that anybody is keeping track) but I believe these functions were written by Dave who is by no means a jr PHP dev. My point is not to cast blame but just to say that everyone makes mistakes sometimes.

Given that a lot of code is copy-pasted as people learn how to e.g. write modules I think having some little utility functions like this help improve code readability and can help you think more clearly while writing -- and the alternative is keeping track of every little detail of PHP weirdness which is pretty tough.

ridz1208 commented:
@johnsaigle I don't make the distinction between new/old and jr/sr developers... everyone makes mistakes, yes including dave !!

But Dave's code got reviewed and merged so there is at least one other developer who also failed to catch the error. The issue in my opinion is not PHP being inconsistent (even if that does not help) but improper management of PRs which is an issue we have been tackling lately and I think we made significand improvements.

Tha said, I'm not THAT opposed to these chnages, it is an improvement... will keep this PR open for further development

johnsaigle commented:
Now that I look this over ain, I don't know that this is actually fixing the problem. Someone unfamiliar with the code would still need to read this over to see if it's `$needle, $haystack` or `$haystack, $needle`.
